File: Angles_real/Angles_real.m
classdef Angles_real < matlab.System & coder.ExternalDependency
        % Angles_real- this block sends its inputs to active X-Plane simulation using shared memory. Note: in order for this block to work, special X-Plane plugin needs to be installed
    
        properties(Access = public)
        
        end
            
        properties(DiscreteState)
    
        end
    
        properties(Access = private)
    
        end
    
        methods(Access = protected)
            function setupImpl(obj)
                    coder.cinclude('Angles_real.h');
                    coder.ceval('setup_Angles_real');
            end
    
            function [pitch,true_heading,roll] = stepImpl(obj)
                pitch = single(0);
                true_heading = single(0);
                roll = single(0);
                coder.cinclude('Angles_real.h');
                coder.ceval('update_Angles_real');
                pitch = coder.ceval('get_pitch');
                true_heading = coder.ceval('get_true_heading');
                roll = coder.ceval('get_roll');
            end
    
            function resetImpl(obj)
                
            end
            
            function releaseImpl(obj)
                coder.cinclude('Angles_real.h');
                coder.ceval('close_Angles_real');
            end
        end
            
        methods (Static)
            function name = getDescriptiveName()
                name = 'Angles_real';
            end
            function b = isSupportedContext(context)
                b = context.isCodeGenTarget('rtw');
            end
            function updateBuildInfo(buildInfo, context)
            % Update the build-time buildInfo
                blockRoot = fileparts(mfilename('fullpath'));
                buildInfo.addIncludePaths({blockRoot});
                buildInfo.addIncludeFiles({'Angles_real.h'});
                buildInfo.addSourcePaths({blockRoot});
                buildInfo.addSourceFiles({'Angles_real.c'});
            end
        end
    end

File: Angles_real/Angles_real.c
#include "Angles_real.h"


void setup_Angles_real()
{
    Angles_real_Mapping = OpenFileMapping(
        FILE_MAP_ALL_ACCESS,   // read/write access
        FALSE,                 // do not inherit the name
        Angles_real_Name);               // name of mapping object
    
    if (Angles_real_Mapping == NULL) {
        Angles_real_isMapped = false;
        return;
    }

    Angles_real_Buf = (LPTSTR)MapViewOfFile(Angles_real_Mapping, // handle to map object
        FILE_MAP_ALL_ACCESS,  // read/write permission
        0,
        0,
        BUF_SIZE);
    
    if (Angles_real_Buf == NULL) {
        Angles_real_isMapped = false;
        return;
    }
    Angles_real_isMapped = true;
}


float get_pitch()
{
    if (Angles_real_isMapped) {
        return Angles_real_struct.pitch;
    } else {
        return 0;
    }
}

float get_true_heading()
{
    if (Angles_real_isMapped) {
        return Angles_real_struct.true_heading;
    } else {
        return 0;
    }
}

float get_roll()
{
    if (Angles_real_isMapped) {
        return Angles_real_struct.roll;
    } else {
        return 0;
    }
}

void update_Angles_real() {
    if (Angles_real_isMapped) {
        memcpy(&Angles_real_struct, Angles_real_Buf, sizeof(Angles_real_struct));
    }
}

void close_Angles_real()
{
    if (Angles_real_isMapped) {
        
        UnmapViewOfFile(Angles_real_Buf);
        CloseHandle(Angles_real_Mapping);
    }
}